mut/redirects/redirect_main.py

```diff
@@ -0,0 +1,396 @@
+    new_rule.version = version
+    new_rule.old_url = old_url_sub
+    new_rule.new_url = new_url_sub
+    
+    # check for symlinks
+    if len(symlinks) > 0:
+        for x in range (len(symlinks)):
+            if version == symlinks[x][1]:
+                r = generate_rule(is_temp, symlinks[x][0], old_url, new_url, True)
+                rules.append(r)
+    
+    return new_rule
+
+def rule_substitue(url_string, version):
+    
+    global definitions
+    
+    # look for strings between { }
+    sub_regex = '{(.*?)}'
+    
+    matches = re.findall(sub_regex, url_string, re.DOTALL)
+    
+    if matches:
+        # loop through each match
+        for x in range(0, len(matches)): 
+            # loop through each definition
+            for y in range(0, len(definitions)):
+                # if the match == the definition key
+                if matches[x] == definitions[y][0]:
+                    # substitute the definition value for the match
+                    if type(definitions[y][1]) is str:
+                        url_string = url_string.replace('${' + matches[x] + '}', definitions[y][1])
+                    
+    url_string = url_string.replace('${version}', version)
+    
+    url_string = url_string.strip()
+    
+    return url_string
+    
+def write_to_file(rules, output):
+    
+    # I'm not sure how we want to handle the file name - maybe it makes sense to have it as a parameter?
+    file = open(output + '/htaccess_test.txt', 'w')
+    
+    for rule in rules:
+        line = 'Redirect '
+        
+        if rule.is_temp:
+            line += '302 '
+        else:
+            line += '301 '
+            
+        line += str(rule.old_url) + ' ' + str(rule.new_url)
+        
+        file.write(line)
+        file.write('\n')
+        
+    file.close()
+
+def parse_source_file(source_path, output):
+    
+    line_num = 0
+    
+    with open(source_path) as file:
+        for line in file:
+            
+            line_num += 1
+            
+            # strip \n from line
+            line = line.strip()
+            
+            # regex to see if we are dealing with a keyword - define, symlink, or raw
+            if re.search('(define|symlink|raw)', line):
+                keyword_split = line.split(':', 1)
+                
+                # define:
+                if keyword_split[0] == 'define':
+                    type_split = keyword_split[1].split(' ')
+                    
+                    key = type_split[1]
+                    
+                    
+                    if len(type_split) > 3:
+                        
+                        value = []
+                        
+                        for x in range (2, len(type_split)):
+                            value.append(type_split[x])
+                            
+                    else:
+                        value = type_split[2]
+                        
+                    definition = [key, value]
```

Comment:
Tuples are much more efficient than lists, if you don't need mutability:

```python
definition = (key, value)
```

Reply:
I believe I do need mutability for when I do the substitutions, so I think I'll have to keep them as lists.